src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1507,32 +1519,62 @@
+                                  ;; ensuring determinism, we respect that ordering here by conj'ing on to the existing
+                                  ;; collection.
+                                  (conj (wrapped-ancestors-fn fact-type) fact-type))))
+        
+        update-roots->facts! (fn [^java.util.Map roots->facts roots-group fact]
+                               (if-let [v (.get roots->facts roots-group)]
+                                 (.add ^java.util.List v fact)
+                                 (.put roots->facts roots-group (doto (java.util.LinkedList.)
+                                                                  (.add fact)))))]
 
     (fn [facts]
-      (for [[fact-type facts] (platform/group-by-seq fact-grouping-fn facts)]
-
-        (if-let [alpha-nodes (get @alpha-map fact-type)]
+      (let [roots->facts (java.util.LinkedHashMap.)]
 
-          ;; If the matching alpha nodes are cached, simply return them.
-          [alpha-nodes facts]
-
-          ;; The alpha nodes weren't cached for the type, so get them now.
-          (let [ancestors (conj (ancestors-fn fact-type) fact-type)
-
-                ;; Get all alpha nodes for all ancestors.  Keep them sorted to maintain
-                ;; deterministic ordering of fact propagation across the network.
-                ;; Alpha nodes do not have a :node-id of their own right now, so sort
-                ;; by the :node-id of their :children.
-                new-nodes (sort-by #(mapv :node-id (:children %))
-                                   (into []
-                                         (comp (map #(get alpha-roots %))
-                                               cat
-                                               (distinct))
-                                         ancestors))]
+        (doseq [fact facts
+                roots-group (fact-type->roots (fact-grouping-fn fact))]
+          (update-roots->facts! roots->facts roots-group fact))
+        
+        (let [return-list (java.util.LinkedList.)]
+          ;; We iterate over the LinkedHashMap manually to avoid potential issues described at http://dev.clojure.org/jira/browse/CLJ-1738
+          ;; where a Java iterator can return the same entry object repeatedly and mutate it after each next() call.  We use mutable lists
+          ;; for performance but wrap them in unmodifiableList to make it clear that the caller is not expected to mutate these lists.
+          ;; Since after this function returns the only reference to the fact lists will be through the unmodifiedList we can depend elsewhere
+          ;; on these lists not changing.  Since the only expected workflow with these lists is to loop through them, not add or remove elements,
+          ;; we don't gain much from using a transient (which can be efficiently converted to a persistent data structure) rather than a mutable type. 
+          (doseq [^java.util.Map$Entry e (.entrySet roots->facts)]
```

Comment:
`doseq` over the entry set will still be subject to the dangers described in the paragraph before since `doseq` is based around "seq"ing over the collection.

To avoid this, you need to explicitly manage a `loop` using manually managed `hasNext()` and `next()` calls.
The up-side of this is that you know you are getting maximum performance too when not going through the seq-abstraction.
```clj
(let [<etc>
      es (.entrySet root->facts)]

  (loop [e (when (.hasNext es)
             (.next es))]

    (.add return-list [(-> e ^AlphaRootsWrapper (.getKey) .roots)
                       (java.util.Collections/unmodifiableList (.getValue e))])

    (when (.hasNext es)
      (recur (.next es)))))
```